File: Kotlin/EsUnAnagrama/EsUnAnagrama.kt
package dev.krispi.fizzbuzz

/*
 * ¿ES UN ANAGRAMA?
 * Dificultad: MEDIA
 *
 * Enunciado: Escribe una función que acepte dos cadenas de texto (String) y devuelva un valor booleano (Bool) que indique si las dos cadenas son anagramas o no.
 * Un anagrama es una palabra o frase que se forma reorganizando todas las letras de otra palabra o frase original.
 * No es necesario verificar si ambas cadenas existen.
 * Es importante destacar que dos cadenas idénticas no se consideran anagramas.
 */

fun sonAnagramas(palabra1: String, palabra2: String): Boolean {
    // Eliminar espacios en blanco y convertir las palabras a minúsculas
    val palabra1SinEspacios = palabra1.replace(" ", "").toLowerCase()
    val palabra2SinEspacios = palabra2.replace(" ", "").toLowerCase()

    // Verificar si tienen la misma longitud
    if (palabra1SinEspacios.length != palabra2SinEspacios.length) {
        return false
    }

    // Convertir las palabras en listas de caracteres y ordenarlas
    val listaCaracteres1 = palabra1SinEspacios.toList().sorted()
    val listaCaracteres2 = palabra2SinEspacios.toList().sorted()

    // Verificar si las listas ordenadas son iguales
    return listaCaracteres1 == listaCaracteres2
}

fun main() {
    val palabra1 = "listen"
    val palabra2 = "silent"
    
    if (sonAnagramas(palabra1, palabra2)) {
        println("$palabra1 y $palabra2 son anagramas.")
    } else {
        println("$palabra1 y $palabra2 no son anagramas.")
    }
}

File: Python/EsUnAnagrama/EsUnAnagrama.py
'''
¿ES UN ANAGRAMA?
Dificultad: MEDIA

Enunciado: Escribe una función que acepte dos cadenas de texto (String) y devuelva un valor booleano (Bool) que indique si las 
dos cadenas son anagramas o no.

Un anagrama es una palabra o frase que se forma reorganizando todas las letras de otra palabra o frase original.
No es necesario verificar si ambas cadenas existen.
Es importante destacar que dos cadenas idénticas no se consideran anagramas.
'''

def son_anagramas(palabra1, palabra2):
    # Eliminar espacios y convertir todo a minúsculas para comparar de manera uniforme
    palabra1 = palabra1.replace(" ", "").lower()
    palabra2 = palabra2.replace(" ", "").lower()

    # Comprobar si las palabras tienen la misma cantidad de letras
    if len(palabra1) != len(palabra2):
        return False

    # Crear diccionarios para contar las letras en ambas palabras
    contador1 = {}
    contador2 = {}

    # Contar las letras en la primera palabra
    for letra in palabra1:
        if letra in contador1:
            contador1[letra] += 1
        else:
            contador1[letra] = 1

    # Contar las letras en la segunda palabra
    for letra in palabra2:
        if letra in contador2:
            contador2[letra] += 1
        else:
            contador2[letra] = 1

    # Comparar si los diccionarios son iguales
    return contador1 == contador2

# Ejemplos de uso
print(son_anagramas("listen", "silent"))  # Debe devolver True
print(son_anagramas("hello", "world"))    # Debe devolver False
print(son_anagramas("anagram", "nagaram")) # Debe devolver True

